File: kits/kotlin/simple/bot/src/main/kotlin/luxaibot/lux/Position.kt
package luxaibot.lux

import java.lang.Error

data class Position(val x: Int, val y: Int) {
    fun isAdjacent(pos: Position): Boolean {
        val dx = x - pos.x
        val dy = y - pos.y
        return if (Math.abs(dx) + Math.abs(dy) > 1) {
            false
        } else true
    }

    fun translate(direction: Direction?, units: Int): Position {
        when (direction) {
            Direction.NORTH -> return Position(x, y - units)
            Direction.EAST -> return Position(x + units, y)
            Direction.SOUTH -> return Position(x, y + units)
            Direction.WEST -> return Position(x - units, y)
            Direction.CENTER -> return Position(x, y)
        }
        throw Error("Did not supply valid direction")
    }

    fun distanceTo(pos: Position): Double {
        return (Math.abs(pos.x - x) + Math.abs(pos.y - y)).toDouble()
    }

    fun directionTo(targetPos: Position): Direction {
        val checkDirections: Array<Direction> =
            arrayOf(Direction.NORTH, Direction.EAST, Direction.SOUTH, Direction.WEST)
        var closestDirection: Direction = Direction.CENTER
        var closestDist = distanceTo(targetPos)
        for (dir in checkDirections) {
            val newpos = translate(dir, 1)
            val dist = targetPos.distanceTo(newpos)
            if (dist < closestDist) {
                closestDist = dist
                closestDirection = dir
            }
        }
        return closestDirection
    }

    override fun toString(): String {
        return "(" + x + ", " + y + ")"
    }
}

File: kits/csharp/simple/Bot/Lux/Position.cs
using System;

namespace Bot.Lux
{
	public class Position
	{
		public int X;
		public int Y;

		public Position(int x, int y)
		{
			X = x;
			Y = y;
		}

		public bool IsAdjacent(Position pos)
		{
			var dx = X - pos.X;
			var dy = Y - pos.Y;

			return Math.Abs(dx) + Math.Abs(dy) <= 1;
		}

		public bool Equals(Position pos)
		{
			return X == pos.X && Y == pos.Y;
		}

		public Position Translate(string direction, int units)
		{
			return direction switch
			{
				Direction.NORTH => new Position(X, Y - units),
				Direction.EAST => new Position(X + units, Y),
				Direction.SOUTH => new Position(X, Y + units),
				Direction.WEST => new Position(X - units, Y),
				Direction.CENTER => new Position(X, Y),
				_ => throw new Exception("Did not supply valid direction")
			};
		}

		public double DistanceTo(Position pos)
		{
			return Math.Abs(pos.X - X) + Math.Abs(pos.Y - Y);
		}
		
		public string DirectionTo(Position targetPos)
		{
			var checkDirections = new []{ Direction.NORTH, Direction.EAST, Direction.SOUTH, Direction.WEST, };
			var closestDirection = Direction.CENTER;
			var closestDist = DistanceTo(targetPos);

			foreach(var dir in checkDirections)
			{
				var newPos = Translate(dir, 1);
				var dist = targetPos.DistanceTo(newPos);
				if (!(dist < closestDist)) continue;
				closestDist = dist;
				closestDirection = dir;
			}

			return closestDirection;
		}

		public override string ToString()
		{
			return "(" + X + ", " + Y + ")";
		}
	}
}
